## 1. What is a neural network? What are the general steps required to build a neural network? 

A neural network is a series of algorithms that can effectively recognize underlying relationships in a set of data through a process that mimics the way the human brain operates. Neural network is a powerful modeling approach that accounts for interactions between data really well.

<b>The general steps required to build a neural network are:</b>

* put together an input layer and define number of input nodes
* a hidden layer or layers and define the number of nodes in these hidden layers
* define the number of ourput nodes; number of output classes you want the neural network to process
* define an activation function for each layer, for example, relu or linear

## 2. Generally, how do you check the performance of a neural network? Why?

To test model's performance, we use the validation data. Because the he model's performance on the training data is not a good indication of how it'd perform on new data. Validation data is the data that's only used to test model performance and nothing else. 

For the validation in deep learning, we use validation split rather than cross-validation, since deep learning is mostly used for larger datasets and cross validations take a lot of computational power. So we can specify the validation split when fitting the model.

## 3. Create a neural network using keras to predict the outcome of either of these datasets: 
Cardiac Arrhythmia: https://archive.ics.uci.edu/ml/datasets/Arrhythmia 
Abalone age: https://archive.ics.uci.edu/ml/datasets/Abalone

In [1]:
import numpy as np
import pandas as pd
import pandas as pd
from keras.layers import Dense
from keras.models import Sequential
from keras.utils.np_utils import to_categorical
from sklearn.preprocessing import StandardScaler

In [2]:
#red the dataset file
abalone_df = pd.read_csv("abalone_edited.csv")

In [3]:
abalone_df.head()

Sex  lengths  Diameter  height  Whole weight  Shucked weight  \
0   M    0.455     0.365   0.095        0.5140          0.2245   
1   M    0.350     0.265   0.090        0.2255          0.0995   
2   F    0.530     0.420   0.135        0.6770          0.2565   
3   M    0.440     0.365   0.125        0.5160          0.2155   
4   I    0.330     0.255   0.080        0.2050          0.0895   

   Viscera weight  Shell weight  Rings  
0          0.1010         0.150     15  
1          0.0485         0.070      7  
2          0.1415         0.210      9  
3          0.1140         0.155     10  
4          0.0395         0.055      7

In [4]:
#checking data types
abalone_df.dtypes

Sex                object
lengths           float64
Diameter          float64
height            float64
Whole weight      float64
Shucked weight    float64
Viscera weight    float64
Shell weight      float64
Rings               int64
dtype: object

In [5]:
#dropping "sex" cilumn
new = abalone_df.drop("Sex", axis=1)

In [6]:
#scaling data
scaler = StandardScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(new), columns = new.columns)

In [7]:
df_scaled.head()

lengths  Diameter    height  Whole weight  Shucked weight  Viscera weight  \
0 -0.574558 -0.432149 -1.064424     -0.641898       -0.607685       -0.726212   
1 -1.448986 -1.439929 -1.183978     -1.230277       -1.170910       -1.205221   
2  0.050033  0.122130 -0.107991     -0.309469       -0.463500       -0.356690   
3 -0.699476 -0.432149 -0.347099     -0.637819       -0.648238       -0.607600   
4 -1.615544 -1.540707 -1.423087     -1.272086       -1.215968       -1.287337   

   Shell weight     Rings  
0     -0.638217  1.571544  
1     -1.212987 -0.910013  
2     -0.207139 -0.289624  
3     -0.602294  0.020571  
4     -1.320757 -0.910013

In [51]:
X = df_scaled.drop('Rings', axis=1)
y = df_scaled['Rings']

In [52]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state=42)

In [53]:
#building a model
model = Sequential()
model.add(Dense(10, activation='relu'))
model.add(Dense(1, activation = 'linear'))

In [54]:
model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae', 'RootMeanSquaredError'])

In [55]:
model.fit(X_train, y_train, batch_size=20, epochs = 50, validation_split = 0.3)

Epoch 1/50
117/117 [==============================] - 1s 3ms/step - loss: 0.7052 - mae: 0.5982 - root_mean_squared_error: 0.8376 - val_loss: 0.5106 - val_mae: 0.5271 - val_root_mean_squared_error: 0.7146
Epoch 2/50
117/117 [==============================] - 0s 3ms/step - loss: 0.5410 - mae: 0.5302 - root_mean_squared_error: 0.7348 - val_loss: 0.4876 - val_mae: 0.5173 - val_root_mean_squared_error: 0.6983
Epoch 3/50
117/117 [==============================] - 0s 2ms/step - loss: 0.5413 - mae: 0.5338 - root_mean_squared_error: 0.7347 - val_loss: 0.4725 - val_mae: 0.4964 - val_root_mean_squared_error: 0.6874
Epoch 4/50
117/117 [==============================] - 0s 2ms/step - loss: 0.5055 - mae: 0.5048 - root_mean_squared_error: 0.7102 - val_loss: 0.4505 - val_mae: 0.4882 - val_root_mean_squared_error: 0.6712
Epoch 5/50
117/117 [==============================] - 0s 2ms/step - loss: 0.4832 - mae: 0.5008 - root_mean_squared_error: 0.6947 - val_loss: 0.4383 - val_mae: 0.4810 - val_root_mean_sq

117/117 [==============================] - 0s 2ms/step - loss: 0.4054 - mae: 0.4590 - root_mean_squared_error: 0.6362 - val_loss: 0.3874 - val_mae: 0.4533 - val_root_mean_squared_error: 0.6224
Epoch 42/50
117/117 [==============================] - 0s 3ms/step - loss: 0.4022 - mae: 0.4593 - root_mean_squared_error: 0.6333 - val_loss: 0.3824 - val_mae: 0.4488 - val_root_mean_squared_error: 0.6184
Epoch 43/50
117/117 [==============================] - 0s 2ms/step - loss: 0.4534 - mae: 0.4842 - root_mean_squared_error: 0.6730 - val_loss: 0.3801 - val_mae: 0.4422 - val_root_mean_squared_error: 0.6165
Epoch 44/50
117/117 [==============================] - 0s 2ms/step - loss: 0.4297 - mae: 0.4688 - root_mean_squared_error: 0.6544 - val_loss: 0.3810 - val_mae: 0.4398 - val_root_mean_squared_error: 0.6172
Epoch 45/50
117/117 [==============================] - 0s 2ms/step - loss: 0.4705 - mae: 0.4888 - root_mean_squared_error: 0.6853 - val_loss: 0.3819 - val_mae: 0.4377 - val_root_mean_squared_e

## 4. Write another algorithm to predict the same result as the previous question using either KNN or logistic regression.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [ ]:
#logostic regression
X = df_scaled.drop('Rings', axis=1) #scaling X
y = new['Rings']

#split data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=50)


model = LogisticRegression(random_state=42)
model.fit(X_train, y_train)
model.score(X_test, y_test)

In [ ]:
from sklearn.metrics import mean_squared_error
y_pred = model.predict(X_test)
mean_squared_error(y_test, y_pred)

## 5. Create a neural network using pytorch to predict the same result as question 3

In [74]:
import torch
import torch.nn as nn
import torch.nn.functional as F #this has activation functions

X = df_scaled.drop('Rings', axis=1)
y = df_scaled['Rings']

#split data into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=50)


# Creating tensors
X_train = torch.FloatTensor(X_train.to_numpy())
X_test = torch.FloatTensor(X_test.to_numpy())

y_train = torch.FloatTensor(y_train.to_numpy())
y_test = torch.FloatTensor(y_test.to_numpy())

print(X_train)

tensor([[ 0.5497,  0.5252,  0.1311,  ...,  0.2687, -0.0191,  0.1521],
        [-0.3664, -0.3818, -0.2275,  ..., -0.5041, -0.2746, -0.5700],
        [ 0.0500,  0.0717,  0.0116,  ..., -0.2202, -0.1195, -0.2431],
        ...,
        [-1.6572, -1.4903, -1.1840,  ..., -1.2362, -1.3193, -1.2705],
        [-0.4496, -0.3314, -0.8253,  ..., -0.8668, -0.8585, -0.6741],
        [ 0.7995,  0.8276,  0.2507,  ...,  0.4557,  1.1077,  0.8346]])


In [75]:
class ANN_Model(nn.Module):
    def __init__(self, input_features=7, hidden1=20, hidden2=20, out_features =2):
        super().__init__()
        self.layer_1_connection = nn.Linear(input_features, hidden1)
        self.layer_2_connection = nn.Linear(hidden1, hidden2)
        self.out = nn.Linear(hidden2, out_features)
    
    def forward(self, x):
        #apply activation functions
        x = F.relu(self.layer_1_connection(x))
        x = F.relu(self.layer_2_connection(x))
        x = self.out(x)
        return x

In [76]:
torch.manual_seed(42)

#instantiate the model
model1 = ANN_Model()

In [77]:
# loss function
loss_function = nn.CrossEntropyLoss()

#optimizer
optimizer = torch.optim.Adam(model1.parameters(), lr = 0.01)

In [78]:
#run model through multiple epochs/iterations
final_loss = []
n_epochs = 100
for epoch in range(n_epochs):
    y_pred = model1.forward(X_train)
    loss = loss_function(y_pred, y_train)
    final_loss.append(loss)
    
    if epoch % 10 == 1:
        print(f'Epoch number: {epoch} with loss: {loss.item()}')
    
    optimizer.zero_grad() #zero the gradient before running backwards propagation
    loss.backward() #for backward propagation 
    optimizer.step() #performs one optimization step each epoch   

RuntimeError: expected scalar type Long but found Float

## 6. Compare the performance of the neural networks to the other model you created. Which performed better? Why do you think that is?

To evaluate the performance of the models I mainly looked at Root mean square error. Based on the abalone dataset, neural networks model performed best having **root mean square error of 0.6280.**
Logistic regression accuracy turned out to be very low for this dataset - 0.27 and root mean square error very high - 6.74.

I think neural network model performed better because it is known to be better in identifying the relationships and interactions between the model's features.